streams/src/test/java/org/apache/kafka/streams/integration/RelaxedNullKeyRequirementJoinTest.java

```diff
@@ -71,11 +71,21 @@ void afterEach() {
     @Test
     void testRelaxedLeftStreamStreamJoin() {
         leftStream
-            .leftJoin(rightStream, JOINER, WINDOW)
+            .leftJoin(rightStream, JOINER, WINDOW_60MS_GRACE_10MS)
             .to(OUT);
         initTopology();
-        left.pipeInput(null, "leftValue", 1);
-        assertEquals(Collections.singletonList(new KeyValue<>(null, "leftValue|null")), out.readKeyValuesToList());
+        left.pipeInput(null, "leftValue1", 1);
+        left.pipeInput(null, "leftValue2", 90);
+        left.pipeInput(null, "lateArrival-Dropped", 19);
+        left.pipeInput(null, "lateArrivalWithinGrace", 20);
+        assertEquals(
+            Arrays.asList(
+                new KeyValue<>(null, "leftValue1|null"),
+                new KeyValue<>(null, "leftValue2|null"),
+                new KeyValue<>(null, "lateArrivalWithinGrace|null")
```

Comment:
Why is this record already in the output? We should not drop it, but it seems we cannot emit it right away either, because we need to wait until the window closes, so would need to pipe one more record with ts=91 to flush out this result?

Reply:
Yes, good point. As is, the KStreamKStreamJoin forwards null-key records directly and not only after the window closes.

Reply:
I wonder how to best solve this.
Multiple null-key records would colide in the 'outerJoinStore' as they potentially all could have the same key of  type 'TimestampedKeyAndJoinSide<K>'.

E.g. for  a left stream

| Key  | value | ts |
|------|-------|----|
| null | a     | 1  |
| null | b     | 1  |
| null | c     | 1  |

We probably would only get to see 'c' even though we would like to see 'a','b' and 'c'?

From the top of my head I see two options to handle this.
- Maintain an additional store just for null-key records where such records wouldn't collide.
- Adjust the 'outerJoinStore' key type `TimestampedKeyAndJoinSide<K>'`( E.g. by adding offset and partition as optional fields. This way null-key records could be distinguished and for keyed records the old behavior can be kept)

Personally I prefer the latter.

Reply:
I just realized that partition and offset are not available in this context...
So, with regards to the second bullet point, probably the only distinguishing feature would be the value itself?

I start to wonder, is it really worth it to add this complexit? We know all three events should eventually be sent downstream anyway. This way we could avoid using additional storage & extending `TimestampedKeyAndJoinSideSerde` with another generic field.

Reply:
Oh. This is `null`-key case. My bad. For `null`-key we know it won't join in the future, so no reason to artificially delay the output. Thanks for pointing it out.